File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/CreatureSizeEnum.java
package ai.ragu.charactermanager.enumeration;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum CreatureSizeEnum implements WeaponTraitEnum {
    SMALL("small"),
    MEDIUM("medium"),
    LARGE("large");

    private final String value;
    private final static Map<String, CreatureSizeEnum> CONSTANTS = new HashMap<>();

    static {
        for (CreatureSizeEnum c: values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    CreatureSizeEnum(String value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return this.value;
    }

    @JsonValue
    public String value() {
        return this.value;
    }

    @JsonCreator
    public static CreatureSizeEnum fromValue(String value) {
        CreatureSizeEnum constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new IllegalArgumentException(value);
        } else {
            return constant;
        }
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/CreatureSizeEnum.kt
package ai.ragu.charactermanager.enumeration

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.util.*

enum class CreatureSizeEnum(private val value: String) : WeaponTraitEnum {
    SMALL("small"),
    MEDIUM("medium"),
    LARGE("large");

    companion object {
        private val CONSTANTS: MutableMap<String, CreatureSizeEnum> = HashMap()

        @JsonCreator
        fun fromValue(value: String?): CreatureSizeEnum {
            val constant = CONSTANTS[value]
            return constant ?: throw IllegalArgumentException(value)
        }

        init {
            for (c in values()) {
                CONSTANTS[c.value] = c
            }
        }
    }

    override fun toString(): String {
        return value
    }

    @JsonValue
    fun value(): String {
        return value
    }

}